### PR TITLE
Rewrite DCE pass

### PR DESCRIPTION
**Pull request by kripken**
The DCE pass is one of the oldest in binaryen, and had quite a lot of
cruft from the changes in unreachability and other stuff in wasm and
binaryen's history. This PR rewrites it from scratch, making it about
1/3 the size.

I noticed this when looking for places to use code autogeneration.
The old version had annoying boilerplate, while the new one avoids
any need for it.

There may be noticeable differences, as the old pass did more than
it needed to. It overlapped with `remove-unused-names` for some
reason I don't remember. The new pass leaves that to the other
pass to do. I added another run of `remove-unused-names` to avoid
noticeable differences in optimized builds, but you can see
differences in the testcases that only run DCE by itself. (The test
differences in this PR are mostly whitespace.)

(The overlap is that if a block ended up not needed, that is, all
branches to it were removed, the old DCE would remove the block.)

This pass is about 15% faster than the old version. However, when
adding another run of `remove-unused-names` the difference
basically vanishes, so this isn't a speedup.

### CODE REVIEW

**Review by tlively — APPROVED** [review]
LGTM!

File: test/passes/dce_all-features.wast [hunk 1]
```diff
@@ -810,4 +810,50 @@
...
+      (i32.const 0)
+      (return (i32.const 1))
+      (unreachable)
+    )
+  )
+  (func $note-loss-of-if-children
+    (block $label$1
+     (if ;; begins unreachable - type never changes - but after the condition
+         ;; becomes unreachable, it will lose the children, which means no more
+         ;; br to the outer block, changing that type.
```
Comment (aheejin) [inline test/passes/dce_all-features.wast:836]: Not sure what changes in the outer block... Doesn't the outer block also begins unreachable, given that it does not have a type attached and its only child (`if`) starts unreachable?

File: test/passes/dce_all-features.wast [hunk 2]
```diff
@@ -810,4 +810,50 @@
...
+  )
+  (func $note-loss-of-non-control-flow-children
+    (block $out
+      (drop
+       (i32.add
+         (block (result i32)
+            (nop)
+            (unreachable)
+         )
+         (br $out) ;; when this is removed as dead, the block becomes unreachable
```
Comment (aheejin) [inline test/passes/dce_all-features.wast:854]: This too.. Doesn't the outer block starts unreachable anyway, given that `(br $out)`'s unreachability propagates up when initially creating the AST?

**Review by aheejin — COMMENTED** [review]
Looks nice and a lot more concise! Some minor questions:

File: src/ir/type-updating.h [hunk 3]
```diff
@@ -147,6 +147,14 @@ struct TypeUpdater
     discoverBreaks(curr, parent ? +1 : -1);
   }
 
+  // Applies a type change to a node, and potentially to its parents.
+  void doTypeChange(Expression* curr, Type type) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/ir/type-updating.h]: Done.

File: src/passes/DeadCodeElimination.cpp [hunk 4]
```diff
@@ -58,514 +60,111 @@ struct DeadCodeElimination
...
-    if (isUnreachable(curr->target)) {
-      auto* block = getModule()->allocator.alloc<Block>();
-      for (auto* operand : curr->operands) {
-        block->list.push_back(drop(operand));
+      // Finally, if there is no need for a concrete type (which is when there
+      // is one marked, but nothing breaks to it, and also the block does not
+      // have a concrete value flowing out) then remove it, which may allow
+      // more reduction.
+      if (block->type.isConcrete() && block->name.is() &&
+          typeUpdater.blockInfos[block->name].numBreaks == 0 &&
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/DeadCodeElimination.cpp]: Good idea, done.

File: src/passes/DeadCodeElimination.cpp [hunk 5]
```diff
@@ -58,514 +60,111 @@ struct DeadCodeElimination
...
+          iff->ifTrue->type == Type::unreachable &&
+          iff->ifFalse->type == Type::unreachable) {
+        typeUpdater.doTypeChange(iff, Type::unreachable);
+      }
+    } else if (auto* loop = curr->dynCast<Loop>()) {
+      // The loop body may have unreachable type if it branches back to the
+      // loop top, for example. The only case we look for here is where we've
+      // already removed the entire body as dead code.
+      if (loop->body->is<Unreachable>()) {
+        replaceCurrent(loop->body);
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/DeadCodeElimination.cpp:159]: RemoveUnusedNames will remove a loop which has no breaks to it, so we can leave it to that. I don't think there is any downside that I can see to doing less here.

File: src/passes/DeadCodeElimination.cpp [hunk 6]
```diff
@@ -58,514 +60,111 @@ struct DeadCodeElimination
...
+      }
+    } else if (curr->is<Try>()) {
+      // TODO
```
[the remark this answers is not in the document]
Comment (kripken) [reply src/passes/DeadCodeElimination.cpp]: thanks, applied.

File: test/passes/dce_all-features.wast (hunk 1 above)
Comment (kripken) [reply test/passes/dce_all-features.wast:847]: No, the block has a `br` to it (showing it can be reached and exited normally), so it begins with type `none`. Same as

```wat
(block $foo ;; type none
  (block $child ;; type unreachable
    (br $foo) ;; this makes $foo reachable
    (unreachable)
  )
)
```

File: test/passes/dce_all-features.wast (hunk 2 above)
Comment (kripken) [reply test/passes/dce_all-features.wast:865]: (same as before I think? but not sure what you mean by `(br $out)'s unreachability propagates up`)

File: test/passes/interesting-pass-mix.txt [hunk 7]
```diff
@@ -41,14 +41,16 @@
   )
   (loop $shape$4$continue
    (call $trivial)
-   (br_if $shape$4$continue
+   (if
     (local.get $0)
+    (br $shape$4$continue)
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/passes/interesting-pass-mix.txt:46]: Yes, RemoveUnsedBrs should turn a simple `if` into a `br_if` when possible.

File: test/wasm2js/br.2asm.js [hunk 8]
```diff
@@ -106,28 +106,34 @@ function asmFunc(global, env) {
  function $13() {
   var $0 = 0, $1_1 = 0, $3_1 = 0;
   block : {
-   $0 = 3;
-   break block;
+   loop_in : while (1) {
```
[the remark this answers is not in the document]
Comment (kripken) [reply test/wasm2js/br.2asm.js:109]: I'm confused, isn't the diff the opposite - new `while`s appear? That's expected as we don't remove a `loop` that is not branched to anymore.

**Review by kripken — COMMENTED** [review]
Thanks for the feedback!

File: test/wasm2js/br.2asm.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (aheejin) [reply test/wasm2js/br.2asm.js:109]: Ah you're right, I was confused. Thanks.

File: test/passes/dce_all-features.wast (hunk 1 above)
Comment (aheejin) [reply test/passes/dce_all-features.wast:847]: I see, thanks. (You mean, in the comment in your code snippet, "this makes $child unreachable"?)

**Review by aheejin — APPROVED** [review]
Thanks!

File: test/passes/dce_all-features.wast (hunk 1 above)
Comment (kripken) [reply test/passes/dce_all-features.wast:847]: Oh, it does both I guess, yes: that one `br` makes `child` unreachable, and also makes `foo` reachable.
